chia-client/src/peer.rs

```diff
@@ -0,0 +1,219 @@
+    NewPeakWallet(NewPeakWallet),
+}
+
+pub struct Peer {
+    sink: Mutex<SplitSink<WebSocket, tungstenite::Message>>,
+    inbound_task: JoinHandle<()>,
+    event_receiver: broadcast::Receiver<PeerEvent>,
+    requests: Requests,
+    nonce: Mutex<u16>,
+}
+
+struct ResponseHandler {
+    requests: Requests,
+    message_id: u16,
+}
+
+impl ResponseHandler {
+    async fn new(
+        requests: Requests,
+        message_id: u16,
+    ) -> (ResponseHandler, oneshot::Receiver<Message>) {
+        let (sender, receiver) = oneshot::channel::<Message>();
+        requests.lock().await.insert(message_id, sender);
+
+        (
+            Self {
+                requests,
+                message_id,
+            },
+            receiver,
+        )
+    }
+}
+
+impl Drop for ResponseHandler {
+    fn drop(&mut self) {
+        let requests = Arc::clone(&self.requests);
+        let message_id = self.message_id;
+
+        tokio::spawn(async move {
+            requests.lock().await.remove(&message_id);
+        });
+    }
+}
+
+impl Peer {
+    pub fn new(ws: WebSocket) -> Self {
+        let (sink, mut stream) = ws.split();
+        let (event_sender, event_receiver) = broadcast::channel(32);
+
+        let requests = Requests::default();
+        let requests_clone = Arc::clone(&requests);
+
+        let inbound_task = tokio::spawn(async move {
+            while let Some(message) = stream.next().await {
+                if let Ok(message) = message {
+                    let bytes = message.into_data();
+                    let cursor = &mut Cursor::new(bytes.as_slice());
+
+                    // Parse the message.
+                    let Ok(message) = Message::parse(cursor) else {
+                        continue;
+                    };
+
+                    if let Some(id) = message.id {
+                        // Send response through oneshot channel if present.
+                        if let Some(request) = requests_clone.lock().await.remove(&id) {
+                            request.send(message).ok();
+                        }
+                    } else {
+                        match message.msg_type {
+                            ProtocolMessageTypes::CoinStateUpdate => {
+                                let cursor = &mut Cursor::new(message.data.as_ref());
+                                if let Ok(body) = CoinStateUpdate::parse(cursor) {
+                                    event_sender.send(PeerEvent::CoinStateUpdate(body)).ok();
+                                }
+                            }
+                            ProtocolMessageTypes::NewPeakWallet => {
+                                let cursor = &mut Cursor::new(message.data.as_ref());
+                                if let Ok(body) = NewPeakWallet::parse(cursor) {
+                                    event_sender.send(PeerEvent::NewPeakWallet(body)).ok();
+                                }
+                            }
+                            _ => {}
```

Comment:
this might warrant some logging, but I recognize this is a proof of concept for now